STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -2,139 +2,158 @@
 {
     // Convert data to uint32_t as needed for the CRC computation
-	uint32_t crcData[size/sizeof(uint32_t)];
+    uint32_t crcData[size/sizeof(uint32_t)];
     memcpy(&crcData, data, size);
 
     // Compute CRC of stored data
     uint32_t crcVal = HAL_CRC_Calculate(hcrc, crcData, sizeof(crcData)/sizeof(uint32_t));
     return crcVal;
 }
 
-// Write data to FLASH_ADDR+indx including CRC.
-void writeToFlashSafe(CRC_HandleTypeDef *hcrc, uint32_t indx, uint32_t size, uint8_t *data)
+/*!
+** @brief Store data + computed CRC of the data in FLASH memory.
+*/
+void writeToFlashCRC(CRC_HandleTypeDef *hcrc, uint32_t flash_address, uint8_t *data, uint32_t size)
 {
     __HAL_RCC_WWDG_CLK_DISABLE();
+    uint32_t flash_sector = getFlashSector(flash_address);
 
     // Erase the sector before write
     HAL_FLASH_Unlock();
-    FLASH_Erase_Sector(FLASH_SECTOR, FLASH_VOLTAGE_RANGE_3);
+    FLASH_Erase_Sector(flash_sector, FLASH_VOLTAGE_RANGE_3);
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
 
     // Compute CRC of data to be saved.
-    uint32_t crcVal = computeCRC(hcrc, size, data);
+    uint32_t crcVal = computeCRC(hcrc, data, size);
 
-	// Append CRC converted to uint8_t to data
+    // Append CRC converted to uint8_t to data
     uint8_t crcData[4] = {0};
-	for (uint8_t i = 0; i < 4; i++)
-	{
-		uint8_t convertedCrcTemp = (crcVal >> 8*i) & 0xFF;
-		//*(data + size + i) = convertedCrcTemp;
-		crcData[i] = convertedCrcTemp;
-	}
-
-	__HAL_RCC_WWDG_CLK_DISABLE();
+    for (uint8_t i = 0; i < 4; i++)
+    {
+        uint8_t convertedCrcTemp = (crcVal >> 8*i) & 0xFF;
+        //*(data + size + i) = convertedCrcTemp;
+        crcData[i] = convertedCrcTemp;
+    }
+
+    __HAL_RCC_WWDG_CLK_DISABLE();
     // Write to sector
     HAL_FLASH_Unlock();
     // Save data supplied as input
     for(uint32_t i=0; i<size; i++)
     {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, FLASH_ADDR+indx+i , data[i]);
+        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]);
     }
 
     // Save CRC value
     for(uint32_t i=0; i<4; i++)
     {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, FLASH_ADDR+indx+size+i , crcData[i]);
+        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+size+i , crcData[i]);
     }
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
 }
 
 
-// Read from flash without checking data integrity if CRC module
-// is not enabled in project.
-void readFromFlashSafe(CRC_HandleTypeDef *hcrc, uint32_t indx, uint32_t size, uint8_t *data)
+/*!
+** @brief Read data from FLASH memory at flash_adress and verify data with stored CRC.
+*/
+void readFromFlashCRC(CRC_HandleTypeDef *hcrc, uint32_t flash_address, uint8_t *data, uint32_t size)
 {
     // Read data including CRC value
     for(uint32_t i=0; i<size; i++)
     {
-        *(data + i) = *( (uint8_t *)(FLASH_ADDR+indx+i) );
+        *(data + i) = *( (uint8_t *)(flash_address+i) );
```

Comment:
This feels a bit "wrong way round".
Wouldn't it be better to compute the CRC on the data held in flash memory first, then only copy it into "data" if the CRC matches?
I guess you might take a performance hit though

Reply:
I agree. This has been bothering me before. I will change it.